#US25 - Ajouter des tags à un article

davidbgk opened the pull request:
ETQ éditrice je veux ajouter des tags aux articles afin de mieux les organiser.

- [x] - peut permettre de renseigner un ou plusieurs tags lors de la rédaction d'un article
- [x] - doit permettre l'autosuggestion de tags existants pour éviter les erreurs d'orthographe
- [x] - ne doit pas permettre d'enregistrer des tags de moins de deux caractères
- [x] - doit enregistrer le tag si celui-ci n'existe pas 
- [x] - doit permettre de voir les tags renseignés pour un article
- [x] - doit signaler les tags déjà sauvegardés
- [x] - doit comporter une animation à l'apparition
- [x] - doit permettre de supprimer un tag

davidbgk committed:
Very basic implementation of tags modelisation

vinyll commented on cafebabel/articles/models.py:
```diff
@@ -2,14 +2,33 @@
 from ..users.models import User
 
 
+class Tag(db.Document, UploadableImageMixin):
```
`UploadableImageMixin` requires the implementation of `get_images_url()` and `get_images_url()`

davidbgk replied:
Done.

vinyll commented on cafebabel/articles/models.py:
```diff
@@ -2,14 +2,33 @@
+        return self.name
+
+    @classmethod
+    def update_slug(cls, sender, document, **kwargs):
```
Would be nice to tests this

davidbgk replied:
Done.

vinyll commented on cafebabel/tests/test_articles.py:
```diff
@@ -5,7 +5,7 @@
 from flask.helpers import get_flashed_messages
 
-from ..articles.models import Article
+from ..articles.models import Article, Tag
```
Add `Tag.drop_collection()` in `pytest_runtest_teardown`

davidbgk replied:
Done.

davidbgk committed:
Post-review fixes

davidbgk committed:
Refactor DB management commands

davidbgk committed:
Add fixtures related to articles

vinyll commented on cafebabel/articles/models.py:
```diff
@@ -17,12 +17,18 @@ class Tag(db.Document, UploadableImageMixin):
+        return current_app.config.get('TAGS_IMAGES_URL')
+
+    def get_images_path(self):
+        return current_app.config.get('TAGS_IMAGES_PATH')
```
I guess it should therefore be [here too](https://github.com/cafebabel/cafebabel.com/blob/master/Makefile#L31).
Nevertheless, I know you were suggestion to ensure that the folders were created in another place, and I'm open to discussion.

davidbgk replied:
Done.

davidbgk committed:
More generic fixture tag for tests

davidbgk commented:
For now, tests are slow because all auth related fixtures are loaded for each test. We have to find a better way :)

davidbgk committed:
Add fixtures for tags

davidbgk committed:
Include tags in article detail page

davidbgk committed:
Ensure deployment takes care of tags + fixtures

davidbgk committed:
Merge branch 'master' into 70-tags-addition

# Conflicts:
#	cafebabel/tests/test_drafts.py
#	config.py

vinyll commented on cafebabel/__init__.py:
```diff
@@ -66,22 +65,22 @@ def register_blueprints(app):
+        relations_fixtures(app)
+
+    @app.cli.command(short_help='Load articles and tags fixtures')
+    def load_data():
```
`load_fixtures()`?

vinyll commented on cafebabel/tests/test_articles.py:
```diff
@@ -284,3 +284,35 @@ def test_article_to_translate_should_have_only_other_links(
+    article2 = Article.objects(tags__in=[tag]).first()
+    assert article2.tags[0].summary == 'summary text'
+    assert tag.slug == 'a-tag'
+    assert str(tag) == 'A tag (en)'
```
Feels like these 2 asserts testing the slug value and the stringification should be in a different test.

vinyll commented on cafebabel/tests/test_articles.py:
```diff
@@ -284,3 +284,35 @@ def test_article_to_translate_should_have_only_other_links(
+    assert Article.objects(tags__in=[tag]).count() == 0
+
+
+def test_article_with_tags(app, tag, article):
```
Something more descsriptive like `test_article_deleting_tag_deletes_references`?

vinyll commented on cafebabel/tests/test_articles.py:
```diff
@@ -284,3 +284,35 @@ def test_article_to_translate_should_have_only_other_links(
+    response = client.get(
+        f'/article/{published_article.slug}-{published_article.id}/')
+    assert response.status_code == HTTPStatus.OK
+    assert 'A tag, Tag two.' in response
```
This will very probably quickly break due to html adaptation.
Maybe you could just check the presence of each text individually?

vinyll commented on cafebabel/tests/test_profile.py:
```diff
@@ -31,36 +31,30 @@ def test_author_profile_has_list_of_published_articles_and_drafts(
-    response = client.post(f'/profile/{user.id}/edit/', data={
-        'editor': 'on',
-    })
+    client.post(f'/profile/{user.id}/edit/', data={'editor': 'on'})
```
the possible values coming from the client side are `'1'` or `''`

vinyll commented on cafebabel/tests/test_profile.py:
```diff
@@ -31,36 +31,30 @@ def test_author_profile_has_list_of_published_articles_and_drafts(
-    response = client.post(f'/profile/{user.id}/edit/', data={
-        'editor': 'on',
-    })
+    client.post(f'/profile/{user.id}/edit/', data={'editor': 'on'})
```
same here

vinyll commented on cafebabel/tests/test_profile.py:
```diff
@@ -22,7 +22,7 @@ def test_user_profile_has_list_of_published_articles_no_draft(client, article):
-    login(client, user.email, 'secret')
+    login(client, user.email, 'password')
     published_article.modify(author=user)
     response = client.get(f'/profile/', follow_redirects=True)
```
`f` is useless here

vinyll commented on cafebabel/tests/utils.py:
```diff
@@ -1,8 +1,10 @@
         email=email,
         password=password
     ), follow_redirects=True)
+    assert b'<form action="/login"' not in response.data, response.data
```
Why not `assert 'action="/login"' not in response` ?

davidbgk replied:
We are out of the context where it works.

vinyll commented on cafebabel/tests/test_users.py:
```diff
@@ -8,34 +8,26 @@
     response = client.get('/profile/', follow_redirects=True)
     assert response.status_code == HTTPStatus.OK
-    assert b'<h1>testy@example.com\'s profile</h1>' in response.data
+    assert b'<h1>user@example.com\'s profile</h1>' in response.data
```
`assert "<h1>user@example.com's profile</h1>" in response`?

davidbgk committed:
Post-review fixes

davidbgk committed:
Naive implementation of tags submission from front

vinyll commented on cafebabel/articles/drafts/views.py:
```diff
@@ -24,8 +25,11 @@ def create():
 
     article = Article()
     authors = User.objects.all()
+    tags = {
```
I'm wondering how many tags we may end up with

vinyll commented on cafebabel/articles/drafts/views.py:
```diff
@@ -37,8 +41,11 @@ def edit(draft_id):
     authors = User.objects.all()
+    tags = {
+        language[0]: Tag.objects(language=language[0])
+        for language in current_app.config['LANGUAGES']}
```
As we have this twice, we may want to prefer writing a _staticmethod_ in Tag.

johangiraud committed:
conditionnal input display

vinyll commented on cafebabel/articles/models.py:
```diff
@@ -115,8 +115,14 @@ def save_from_request(self, request):
         for field, value in data.items():
-            setattr(self, field, value)
+            if field.startswith('tag-') and value:
+                tag = Tag.objects.get(name=value)
```
Shouldn't you filter the language here?

johangiraud committed:
add js file + wip display more input

vinyll commented on cafebabel/templates/articles/macros.html:
```diff
@@ -15,3 +15,19 @@
+    <input list=tags name=tag-4 {% if article and article.tags and article.tags[3] %}value={{ article.tags[3].name }}{% endif %}>
+    <input list=tags name=tag-5 {% if article and article.tags and article.tags[4] %}value={{ article.tags[4].name }}{% endif %}>
+    <datalist id=tags>
+      {% for tag in tags['en'] %}
```
Why filtering for English tags instead of the article's language?

johangiraud committed:
style tag input button

vinyll commented on cafebabel/templates/articles/macros.html:
```diff
@@ -17,13 +17,16 @@
+    {% if article and article.tags %}
+    {% set count = 1 %}
+      {% for tag in tags['en'] %}
+        <button class=remove>+</button><input list=tags name=tag-{{ count }} value={{ article.tags.name }}>
```
you may need the quotes for the value here at it may contain space or other special character

vinyll replied:
From a code point of view – without the interface – it looks surprising to have a "+" with a class _remove_

vinyll commented on cafebabel/templates/articles/macros.html:
```diff
@@ -17,13 +17,16 @@
-    <input list=tags name=tag-4 {% if article and article.tags and article.tags[3] %}value={{ article.tags[3].name }}{% endif %}>
-    <input list=tags name=tag-5 {% if article and article.tags and article.tags[4] %}value={{ article.tags[4].name }}{% endif %}>
+    <div class=tag-container><input name=tag-new><button class=add></button></div>
+    {% if article and article.tags %}
```
`{% if … article.tags %}`: does it still work when creating an article or editing an article with no tag, or will these fields not appear at all?

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -0,0 +1,6 @@
+const tagButtonAdd = document.querySelector('.tags button.add')
+const tagButtonRemove = document.querySelector('.tags button.remove')
+tagButtonAdd.addEventListener('click', (event) => {
```
I see this is stopping the native behavior, but it seems like it's doing nothing instead.
Ane what is the default behavior of a button anyways?

davidbgk commented on cafebabel/templates/articles/macros.html:
```diff
@@ -15,3 +15,22 @@
+    <label for=tag-1>Tags</label>
+    <div class=tag-container><input name=tag-new><button class=add>+</button></div>
+    {% if article and article.tags %}
+    {% set count = 1 %}
```
I think you are looking for [`loop.index`](http://jinja.pocoo.org/docs/2.10/templates/#for) 😉

Commit:
Get tag's value

Commit:
Add list for tags

Commit:
Add a new tag

Commit:
Remove example

Commit:
Don't append duplicates tag

Commit:
Create the tags list instead of append each tag

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -1,6 +1,28 @@
+const tagsList = document.querySelector('.tags .tagsList')
+
+const Tag = {
+  count: () => tagsList.querySelectorAll('li').length,
```
I think a real method would seem more appropriate
```
count () {
  return tagsList.querySelectorAll('li').length
}
```

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -1,6 +1,28 @@
+
+const Tag = {
+  count: () => tagsList.querySelectorAll('li').length,
+  createTag(tagValue) {
```
missing space before `(`

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -1,6 +1,28 @@
+
+    return li
+  },
+  tagValues: () => Array.from(tagsList.querySelectorAll('li')).map(li => li.innerText)
```
same here

vinyll commented on cafebabel/templates/articles/macros.html:
```diff
@@ -20,13 +20,16 @@
-    {% set count = 1 %}
+    <ul class=tagsList>
+      {% set count = 1 %}
       {% for tag in tags['en'] %}
```
Still wondering why we take English only here

vinyll commented on cafebabel/templates/articles/macros.html:
```diff
@@ -20,13 +20,16 @@
     <div class=tag-container><input name=tag-new><button class=add>+</button></div>
-    {% if article and article.tags %}
-    {% set count = 1 %}
+    <ul class=tagsList>
```
We haven't used camelCased class names so far but hyphen separated.
Let's try to be consistent

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -2,5 +2,10 @@ const tagButtonAdd = document.querySelector('.tags button.add')
+  if (!tag) {
+    return
+  }
+  console.log('Tag', tag)
```
You may remove this

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -2,5 +2,10 @@ const tagButtonAdd = document.querySelector('.tags button.add')
 tagButtonAdd.addEventListener('click', (event) => {
   event.preventDefault()
+  const tag = event.target.previousElementSibling.value
+  if (!tag) {
```
A online liner is appropriate here: `if (!tag) return`

vinyll commented on cafebabel/templates/articles/macros.html:
```diff
@@ -20,13 +20,14 @@
-    {% if article and article.tags %}
-    {% set count = 1 %}
+    <ul class=tagsList>
+      <li>Machin<input type=hidden list=tags name=tag-0 value=machin></li>
```
`value=machin` 🙃

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -13,15 +13,16 @@ const Tag = {
   event.preventDefault()
   const tagValue = event.target.previousElementSibling.value
-  if (!tagValue) return
+  if (!tagValue || Tag.tagValues().includes(tagValue)) return
```
Just a suggestion: because you have created a Tag object, it may be interesting to create a method to check `Tag.contains(tagValue)`.

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -16,13 +16,21 @@ const Tag = {
   },
-  tagValues: () => Array.from(tagsList.querySelectorAll('li')).map(li => li.innerText)
+  tagValues: () => Array.from(tagsList.querySelectorAll('li')).map(li => li.innerText),
+  createTagsList: (container, tagValues) => {
```
`createTagsList (container, tagValues) {`

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -16,13 +16,21 @@ const Tag = {
-  tagsList.appendChild(Tag.createTag(tagValue))
+  const tag = event.target.previousElementSibling.value
+  if (!tag || Tag.tagValues().includes(tag)) return
+  const tags = Tag.tagValues().concat(tag)
```
Same for creating a method here. It could be `Tag.add(tag)`

johangiraud committed:
Refactor with class

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -1,36 +1,48 @@
-const tagButtonRemove = document.querySelector('.tags button.remove')
-const tagsList = document.querySelector('.tags .tagsList')
+class Tags {
+  constructor(domContext) {
```
the convention that has been taken in this project is to have a space after the name when declaring a function name. Let's keep this uniform.

pointbar replied:
@vinyll - I think I'm missing something here.
Do you ask to add a space between function name and args ?
Like this:
 constructor (domContext) - instead of: constructor(domContext)

davidbgk replied:
It is now handled by prettierjs

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -1,36 +1,48 @@
+    this.values = Array.from(this.list.querySelectorAll('li')).map(li => li.innerText)
+    this.submission = domContext.querySelector('input[name=tag-new]').value
+  }
+  checkSubmission() {
```
same

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -1,36 +1,48 @@
+  checkSubmission() {
+    return this.submission && ! this._isTag(this.submission)
+  }
+  display() {
```
and here

vinyll replied:
I'd suggest calling it `render` as it seems to become more of a convention.
But this is subjective, up to you ;)

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -1,36 +1,48 @@
-const Tag = {
-  count: () => tagsList.querySelectorAll('li').length,
-  createTag(tagValue) {
+  _isTag(tag) {
```
again

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -1,36 +1,48 @@
+  _isTag(tag) {
+    return this.values.includes(tag)
+  }
+  _createTagsList(container, tagValues) {
```
here too

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -1,36 +1,48 @@
+
+    return container
+  }
+  _createTag(tagValue, index) {
```
and here

johangiraud committed:
Add remove action for tags

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -2,17 +2,16 @@ class Tags {
   }
-  checkSubmission() {
-    return this.submission && ! this._isTag(this.submission)
+  checkSubmission(query) {
```
need a space after function name

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -31,18 +30,35 @@ class Tags {
+  event.target.previousSibling.value = ''
 })
 
+function addTagsRemoveListener() {
```
add a space after function name

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -31,18 +30,35 @@ class Tags {
+  }))
+}
+
+function displayTags(tags) {
```
same here

johangiraud committed:
styles add tags

davidbgk committed:
Move tags to their dedicated folder

davidbgk committed:
Basic implementation of tag suggestion API

johangiraud committed:
wip display transition fadein

johangiraud committed:
Merge branch '70-tags-addition' of github.com:cafebabel/cafebabel.com into 70-tags-addition

johangiraud committed:
add container for comming tags

johangiraud committed:
Wip: First tries with the completion api

davidbgk committed:
Use request from utils instead of fetch

davidbgk committed:
Switch to a more explicit URL for tag suggestion

davidbgk committed:
Use datalist to propose API suggestions to the user

davidbgk commented on cafebabel/templates/articles/macros.html:
```diff
@@ -15,3 +15,12 @@
+{%- macro tags_fields(tags, article=None) -%}
+  <div class=tags>
+    <label for=tag-1>Tags</label>
+    <ul class=tags-list></ul>
```
@cafebabel-webmaster I wasn't sure of the purpose of that list, maybe useless now we have the datalist for suggestion?

davidbgk committed:
Deal with locale aware tag suggestion

davidbgk committed:
Allow to submit new/unknown tags on save

davidbgk committed:
Activate tags for article edition

davidbgk committed:
Fix redirection to draft if published changed status

davidbgk committed:
Fix indentation in template

davidbgk committed:
Only published articles use that edition template

davidbgk committed:
Switch to prettier JS for consistency (see detail)

Configuration (for SublimeText):

{
    "auto_format_on_save": true,
    "prettier_options": {
        "singleQuote": true,
        "semi": false
    }
}

davidbgk committed:
Load existing tags

davidbgk commented on cafebabel/templates/articles/macros.html:
```diff
@@ -15,3 +15,21 @@
+{%- macro tags_fields(article=None) -%}
+  <div class=tags>
+    <label for=tag-1>Tags</label>
+    <ul class=tags-list>
```
Here we use `.tags-list` for style and `.tags ul` on the JS side. We have to be consistent @pointbar @cafebabel-webmaster

davidbgk committed:
Use a custom QuerySet for tags to get_or_create

Commit:
Change css selector: class instead of tag

Commit:
Test prettier with noise.js

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -89,6 +89,6 @@ window.addEventListener('load', addTagsRemoveListener)
   const container = tags.list.cloneNode(false)
   tags.list.replaceWith(tags._createTagsList(container, tags.values))
-  document.querySelector('.tags ul li:last-child').classList.add('fadeIn')
+  document.querySelector('.tags .tags-list li:last-child').classList.add('fadeIn')
```
Not sure why adding yet another class within such a specific `.tags` class

pointbar replied:
@vinyll - @johan creates another -ul- for completion, is your comment stays relevant with this new context?

johan replied:
Johan != @johan on github, FYI, @pointbar. :) / disables notifications.

Commit:
Highlight existing translation

Commit:
Fix typo

Commit:
Change label's language to abreviation

johangiraud committed:
styles autocomplete tag + display input value instead of li innerhtml

davidbgk commented on cafebabel/templates/articles/macros.html:
```diff
@@ -15,3 +15,21 @@
+      {% endif %}
+    </ul>
+    <div class=tag-container><input name=tag-new autocomplete=off list=tags-suggestions><button class=add>+</button></div>
+    <ul id=tags-suggestions class=inactive></ul>
```
@cafebabel-webmaster changing that `datalist` element to a `ul` probably broke the native autocomplete from the browser, not sure that is something we want? https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist

johangiraud replied:
I couldn't style the datalist elements. That's why I switched for `ul` and `li`. we can talk about it later

johangiraud committed:
fadein & fadeout tags when adding and removing

davidbgk committed:
US26 - Page detail for tags, fix #126

davidbgk committed:
Merge branch '70-tags-addition' into 126-tags-browsability

davidbgk committed:
Merge branch 'master' into 70-tags-addition

# Conflicts:
#	config.py

davidbgk committed:
Merge branch '70-tags-addition' into 126-tags-browsability

johangiraud committed:
Move fade effects into class

johangiraud committed:
Refactor: move function to class

davidbgk commented on cafebabel/static/js/edit-article.js:
```diff
@@ -0,0 +1,146 @@
+    )
+  }
+  static addTagsRemoveListener() {
+    const tagsButtonRemove = document.querySelectorAll('.tags .tags-list li a')
```
Reuse `this.list` here?

davidbgk commented on cafebabel/static/js/edit-article.js:
```diff
@@ -0,0 +1,146 @@
+    input.name = `tag-${index + 1}`
+    input.setAttribute('list', 'tags')
+    input.value = tagValue
+    li.append(tagValue)
```
I think you can do `li.append(tagValue, input, a)`

davidbgk commented on cafebabel/static/js/edit-article.js:
```diff
@@ -0,0 +1,146 @@
+class Effect {
+  static fadeIn(element) {
+    element.style.opacity = 0
+    ;(function fade() {
```
Out of curiosity, why IIFEs here and below?

pointbar replied:
Not fine with the old version

davidbgk committed:
Merge branch '70-tags-addition' into 126-tags-browsability

Commit:
Merge branch '70-tags-addition' into 145-highlight-translations

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -27,68 +39,108 @@ class Tags {
     input.value = tagValue
     li.append(tagValue)
     li.append(input)
+    li.append(a)
```
`a` seems like blank with no attribute. Is it normal?

We could also consider a JSX-style for replacing the ~10 lines above – which might be more readable:
```js
return `
  <li>
    <input type=hidden list=tags name=tag-${index + 1} value="${tagValue}">
    <a></a>
  </li>
`
```

pointbar replied:
We don't find any solution to attach event on the pseudo-element ::after.
After tries we add an `a` to attach a -remove link-

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -27,68 +39,108 @@ class Tags {
-      .then(response => response.forEach(handleSuggestion))
-      .catch(console.error.bind(console))
-  })
+class Listener {
```
This seems very specific to tags while the class sounds very generic.
Why not going for a name like `TagEventListener`?

vinyll commented on cafebabel/static/js/edit-article.js:
```diff
@@ -27,68 +39,108 @@ class Tags {
+
+window.addEventListener('load', () => {
+  const tags = new Tags(document.querySelector('.tags'))
+  Tags.addTagsRemoveListener()
```
This `addTagsRemoveListener` method should probably belong to the currently called `Listener` class?

Commit:
Merge pull request #150 from cafebabel/145-highlight-translations

145 - highlight existing translation

Commit:
Add js file for create article

Commit:
Refactor: change static method with method

Commit:
Refactor: move tomultiple arguments instead of multiple instructions

davidbgk commented on cafebabel/static/js/edit-article.js:
```diff
@@ -142,5 +138,5 @@ Listener.keyup()
 
 window.addEventListener('load', () => {
   const tags = new Tags(document.querySelector('.tags'))
```
If I understand correctly the constructor of `Tags` doesn't take any parameter now?

davidbgk commented on cafebabel/templates/articles/drafts/create.html:
```diff
@@ -1,6 +1,8 @@
 {% extends "base.html" %}
 {% import "articles/macros.html" as macros without context %}
 
+{% block extra_js %}<script src={{ url_for('static', filename='js/edit-article.js') }}></script>{% endblock %}
```
Maybe we should better name it `articles-tags.js`, it's kind of weird to have an edit file in the create template 😉

Commit:
Refactor: fadeIn fadeOut methods

davidbgk committed:
Deal with significant extra space, fix #151

It’s not really satisfying but it’s better than the previous one-liner. There is maybe a better solution using CSS.

davidbgk commented on cafebabel/static/js/edit-article.js:
```diff
@@ -0,0 +1,147 @@
+  _createTag(tagValue, index) {
+    const li = document.createElement('li')
+    const input = document.createElement('input')
+    const a = document.createElement('a')
```
I second @vinyll here, semantically it's closer to a `button` element. Thoughts?

davidbgk commented on cafebabel/static/js/edit-article.js:
```diff
@@ -0,0 +1,147 @@
+      const tags = new Tags()
+      if (tags.checkSubmission(submission)) tags.add(submission)
+      tags.display()
+      Effect.fadeIn(document.querySelector('.tags .tags-list li:last-child'))
```
Reuse `tags.list`?

davidbgk commented on cafebabel/static/js/edit-article.js:
```diff
@@ -0,0 +1,147 @@
+      const tags = new Tags()
+      const tagElement = event.target.parentNode
+      tags.remove(tagElement.innerText)
+      Effect.fadeOut(tagElement).then(response => tags.display())
```
I think you don't have a `response` here given that you don't `resolve()` any value, I would just do `.then(tags.display)`

Commit:
Refactor: Use templating instead of js DOM functions

Commit:
Merge branch '70-tags-addition' of github.com:cafebabel/cafebabel.com into 70-tags-addition

Commit:
Refactor: rename class

Commit:
Refactor: move removeEvents() in an other class

Commit:
Refactor: remove unnecessary parameter

Commit:
Refactor: rename js file

davidbgk commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -0,0 +1,150 @@
+  _createTag(tagValue, index) {
+    return `
+      <li>
+        ${tagValue}
```
Bad news, with that implementation [extra space bug](https://github.com/cafebabel/cafebabel.com/issues/151) stroked again, this time on the JS side 🙃

johangiraud replied:
We wrote it as one only line, and it fixes it.

davidbgk replied:
Where/when? The current implementation still adds it.

davidbgk committed:
Merge branch '70-tags-addition' into 126-tags-browsability

davidbgk committed:
US22 - Tags administration, fix #30

davidbgk committed:
Fix tests after PR#150, CI check failed hard!

https://github.com/cafebabel/cafebabel.com/pull/150

davidbgk committed:
Merge branch '70-tags-addition' into 126-tags-browsability

davidbgk committed:
Tag has a name and not a title, copy-pasta failure

davidbgk committed:
Merge branch '126-tags-browsability' into 30-tags-administration

davidbgk committed:
Refactoring of the JS, iso-functional

davidbgk committed:
Merge branch '70-tags-addition' into 126-tags-browsability

davidbgk committed:
Merge branch '126-tags-browsability' into 30-tags-administration

vinyll committed:
tag editing now takes the current language

johangiraud committed:
replace remove tag <a> by <button>

Commit:
Refactor: group querySelector

Commit:
Refactor: fix remove tag

Commit:
Refactor: cleanning -add- and -remove- listener

vinyll commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -2,13 +2,15 @@ class Tags {
-    this.buttonAdd = context.querySelector('input[name=tag-new]')
+    this.fieldAdd = context.querySelector('input[name=tag-new]')
+    this.buttonAdd = context.querySelector('button.add')
+    this.removeButtons = context.querySelectorAll('.tags-list li button')
```
`querySelectorAll` returns a _NodeList_.
In some browsers, `forEach` is not implemented on a NodeList.
It may be safer to convert it into an _array_: `this.removeButtons = Array.from(context.querySelectorAll('.tags-list li button'))`

johangiraud committed:
styles detail tag page + parallax on picture

johangiraud committed:
clean up test css values

Commit:
wip: pause

johangiraud committed:
wip styles edit tag page

johangiraud committed:
cleaning after review + change to more precise js file name

johangiraud committed:
styles article list

Commit:
Handle suggestions

davidbgk commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -52,7 +80,7 @@ class TagEffect {
       }
     }
-
+    console.log('TOC FADE')
```
To be removed

johangiraud committed:
add specific metas for tag page

davidbgk commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -0,0 +1,166 @@
+      const submission = event.target.value
+      if (submission.length < 3) return
+      const tags = new Tags()
+      request(`/article/tag/suggest/?language=en&terms=${submission}`)
```
Why removing the dynamic language choice?!

pointbar replied:
@davidbgk - sorry, I misunderstood something about languages

davidbgk committed:
Minor test refactoring

Commit:
Render suggestions

johangiraud committed:
RE styles detail tag page + parallax on picture

johangiraud committed:
RE wip styles edit tag page

johangiraud committed:
RE cleaning after review + change to more precise js file name

johangiraud committed:
RE styles article list

johangiraud committed:
RE add specific metas for tag page

johangiraud committed:
comment edit tag page button

johangiraud committed:
REMOVE add specific metas for tag page

johangiraud committed:
REMOVE styles article list

johangiraud committed:
REMOVE cleaning after review + change to more precise js file name

johangiraud committed:
REMOVE wip styles edit tag page

johangiraud committed:
REMOVE styles detail tag page + parallax on picture

johangiraud committed:
update js file name

johangiraud committed:
correction js file name in template

johangiraud committed:
removing meta block and edit button

johangiraud committed:
put back edition button

Commit:
Link Tag class with real DOM

Commit:
Merge branch '70-tags-addition' of github.com:cafebabel/cafebabel.com into 70-tags-addition

davidbgk committed:
Remove link to edition (not par of that branch)

davidbgk committed:
Merge branch '70-tags-addition' into 126-tags-browsability

# Conflicts:
#	cafebabel/tests/test_tags.py

davidbgk committed:
Merge branch '126-tags-browsability' into 30-tags-administration

# Conflicts:
#	cafebabel/templates/articles/tags/detail.html

Commit:
Adapt language in query

davidbgk committed:
Prevent the edition of tag name and language

Indirectly the slug becomes permanent too which solves a couple of nasty side-cases (redirections and so on)

davidbgk committed:
Force the uniqueness of tag slug+language at DB level

Not sure if it works with different case though.

davidbgk committed:
Merge branch '70-tags-addition' into 126-tags-browsability

davidbgk committed:
Merge branch '126-tags-browsability' into 30-tags-administration

davidbgk commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -0,0 +1,180 @@
+    return this.context.querySelector('input[name=tag-new]')
+  }
+  get removeButtons() {
+    return this.context.querySelectorAll('.tags-list li button')
```
Reuse `this.list()`?

pointbar replied:
Yes, but`this.list`;)

davidbgk commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -0,0 +1,180 @@
+    const inputNewTag = document.querySelector('.tags input[name=tag-new]')
+    inputNewTag.addEventListener('keypress', event => {
+      /* Return if arrow up, arrow down or enter are pressed */
+      if (event.keyCode == 13) {
```
Why the recent split with arrows detection?

pointbar replied:
I try to resolve the problem with -keycode: 13- 
Can I remove this case, this `addEventListener`never captures -return key-

vinyll replied:
It seems obvious to you, but _why_ do we need to block them?

davidbgk commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -0,0 +1,180 @@
+      request(
+        `/article/tag/suggest/?language=${tags.language}&terms=${submission}`
+      )
+        .then(tagsApi => tags.handleSuggestion(tagsApi))
```
Can be simplified as `.then(tags.handleSuggestion)`

pointbar replied:
@davidbgk - I feel like I'm losing `this` and the argument?

davidbgk replied:
Ah! Maybe not that simple 👌

Commit:
Refactor: priorise specific context to select element

Commit:
Fix: remove extra space

Commit:
Merge branch '70-tags-addition' of github.com:cafebabel/cafebabel.com into 70-tags-addition

vinyll committed:
Tag detail filters language

Commit:
Refactor: move request into Tags class

Commit:
Refactore: declare Tags once

vinyll commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -69,13 +69,11 @@ class Tags {
-      <li>
+    return `<li>
         ${tagValue}
         <input name="tag-${++index}" list="tags" value=${tagValue} type="hidden">
```
I suppose you need quotes nowhere but on the `value` attribute as it's the one that may have spaces or special chars 😅

vinyll commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -40,14 +40,18 @@ class Tags {
+  handleSuggestion(submission) {
+    _request(submission)
+      .then(tagsApi => this._createSuggestionList(tagsApi))
+      .catch(console.error.bind(console))
```
Should we alert the user if an error occurs?

vinyll commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -40,14 +40,18 @@ class Tags {
+    )
+  }
+  _isTagSaved(tag) {
+    return
```
🙃

pointbar replied:
I am writing on

vinyll commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -174,12 +169,12 @@ class TagEventListener {
       if (event.keyCode == 40 || event.keyCode == 38) return
```
Could you code-comment what these keyCodes are?
Eventually also _why_ to prevent key up and down as it's not obvious from here.

Commit:
Refactor: remove quote

Commit:
Check if tags is saved

davidbgk committed:
Merge branch '70-tags-addition' into 126-tags-browsability

davidbgk committed:
Ensure fallback to a default language

davidbgk committed:
Only link to tags if article is published

davidbgk committed:
Link in article cards only on the title, fix #155

davidbgk committed:
Merge branch '126-tags-browsability' into 30-tags-administration

# Conflicts:
#	cafebabel/articles/tags/views.py

vinyll commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -48,8 +48,10 @@ class Tags {
-    return
+  _isTagSaved(submission) {
+    return this._request(submission).then(
+      tagsApi => !!tagsApi.filter(tagApi => tagApi.name === submission).length
```
Unless I missunderstood, it would be more straightforward to write:
```js
tagsApi => tagsApi.some(tag => tag.name === submission)
```

johangiraud committed:
style tags and language selector in article detail

Commit:
Highlight saved tags

Commit:
Merge branch '70-tags-addition' of github.com:cafebabel/cafebabel.com into 70-tags-addition

Commit:
Refactor: change -filter- with -some-

vinyll commented on cafebabel/templates/articles/detail.html:
```diff
@@ -34,11 +34,10 @@
+    <p class=tags-list>
       {% for tag in article.tags -%}
-        {{ tag.name }}{% if loop.last %}.{% else %}, {% endif %}
+        {#<a href={{ url_for('tags.detail', slug=tag.slug) }}>{{ tag.name }}</a>#} <!-- next PR -->
```
`<!-- next PR -->`? 🙃

johangiraud replied:
oui ça sera la ligne à décommenter dans la prochaine PR

Commit:
Remove -enter- key in eventListener

johangiraud committed:
check if article has tag

johangiraud committed:
Merge branch '70-tags-addition' of github.com:cafebabel/cafebabel.com into 70-tags-addition

davidbgk commented on cafebabel/templates/articles/detail.html:
```diff
@@ -34,13 +34,14 @@
-        {{ tag.name }}{% if loop.last %}.{% else %}, {% endif %}
-      {%- endfor %}
-    </p>
+    {% if article.tags|length != 0 %}
```
`if article.tags` is enough.

johangiraud replied:
ok

davidbgk commented on cafebabel/templates/articles/detail.html:
```diff
@@ -34,13 +34,14 @@
+    {% if article.tags|length != 0 %}
+      <p class=tags-list>
+        {% for tag in article.tags -%}
+          {#<a href={{ url_for('tags.detail', slug=tag.slug) }}>{{ tag.name }}</a>#} <!-- next PR -->
```
It has nothing to do here.

davidbgk commented on cafebabel/templates/articles/detail.html:
```diff
@@ -34,13 +34,14 @@
+      <p class=tags-list>
+        {% for tag in article.tags -%}
+          {#<a href={{ url_for('tags.detail', slug=tag.slug) }}>{{ tag.name }}</a>#} <!-- next PR -->
+          {{ tag.name }}
```
Please get back to the previous version: `{{ tag.name }}{% if loop.last %}.{% else %}, {% endif %}` or change tests (or explain why you did that in the commit message)

johangiraud replied:
I want to have that in the end: ![capture d ecran 2018-01-18 a 16 30 02](https://user-images.githubusercontent.com/32518627/35105809-e500f8f6-fc6c-11e7-81ac-aa3c246fcdfa.png) so I don't need commas or dot

davidbgk replied:
I know, that's just not the right branch/PR.

johangiraud committed:
set parallax effect to half of image size

Commit:
Fix: remove for chrome

Commit:
Fix: fix error with last tag removing

Commit:
Merge branch '70-tags-addition' of github.com:cafebabel/cafebabel.com into 70-tags-addition

johangiraud committed:
correction after review, delete changes

johangiraud committed:
add article link on thumbnail

johangiraud committed:
Merge branch '126-tags-browsability' into 30-tags-administration

johangiraud committed:
move edit button below tag title

johangiraud committed:
Merge branch '30-tags-administration' of github.com:cafebabel/cafebabel.com into 30-tags-administration

johangiraud committed:
Merge pull request #153 from cafebabel/30-tags-administration

US22 - Tags administration, fix #30

johangiraud committed:
Merge branch '70-tags-addition' into 126-tags-browsability

Commit:
Submit tag with enter key

Commit:
Merge branch '70-tags-addition' of github.com:cafebabel/cafebabel.com into 70-tags-addition

vinyll committed:
tag articles are listed sorted by publication date

vinyll committed:
tag summary is optional

vinyll committed:
restored link on tag for articles and tag display for drafts

davidbgk committed:
Merge branch '70-tags-addition' into 126-tags-browsability

johangiraud committed:
style tag page edit + tag display in article detail

johangiraud committed:
minor detail article corrections

vinyll commented on cafebabel/static/js/articles-tags.js:
```diff
@@ -171,8 +170,10 @@ class TagEventListener {
+      event.preventDefault()
       /* Return if arrow up, arrow down are pressed */
-      if (event.keyCode == 40 || event.keyCode == 38) return
+      if (event.keyCode == 13 || event.keyCode == 40 || event.keyCode == 38)
```
Arrow up, down… and enter?
It would be nice to tell _why_ this hack needs to be processed in very short.

vinyll replied:
Comment should be updated accordingly

Commit:
Fix: update comment

johangiraud committed:
remove useless console.log

Commit:
Remove unnecessaries keys capture

johangiraud committed:
Merge pull request #152 from cafebabel/126-tags-browsability

US26 - Page detail for tags, fix #126

vinyll committed:
fixed test for article tag display with no comma nor point

vinyll committed:
attempting to fix tags test FileNotFound on Travis

vinyll committed:
fixed test tags folder missing

johangiraud committed:
Merge branch 'master' into 70-tags-addition
